test/unit/scan/test_folder_traversal.py

```diff
@@ -0,0 +1,193 @@
+        (d2 / "file3.txt").write_text("content3")
+        (d2 / "file4.txt").write_text("content4")
+
+        folder = LocalFolder(str(tmp_path))
+        local_paths = folder.all_files(reporter=MagicMock())
+        absolute_paths = [path.absolute_path for path in list(local_paths)]
+
+        assert absolute_paths == [
+            fix_windows_path_limit(str(d1 / "file1.txt")),
+            fix_windows_path_limit(str(d1 / "file2.txt")),
+            fix_windows_path_limit(str(d2 / "file3.txt")),
+            fix_windows_path_limit(str(d2 / "file4.txt")),
+        ]
+
+    @pytest.mark.skipif(
+        platform.system() == 'Windows' and platform.python_implementation() == 'PyPy',
+        reason="Symlinks not supported on PyPy/Windows"
+    )
+    def test_folder_with_symlink_to_file(self, tmp_path):
+
+        # Create a directory structure below with initial scannig point at tmp_path:
+        # tmp_path
+        # ├── dir
+        # │   └── file.txt
+        # └── symlink_file.txt -> dir/file.txt
+
+        (tmp_path / "dir").mkdir()
+
+        file = tmp_path / "dir" / "file.txt"
+        file.write_text("content")
+
+        symlink_file = tmp_path / "symlink_file.txt"
+        symlink_file.symlink_to(file)
+
+        folder = LocalFolder(str(tmp_path))
+        local_paths = folder.all_files(reporter=MagicMock())
+
+        absolute_paths = [path.absolute_path for path in list(local_paths)]
+
+        assert absolute_paths == [
+            fix_windows_path_limit(str(file)),
+            fix_windows_path_limit(str(symlink_file))
+        ]
+
+    @pytest.mark.skipif(
+        platform.system() == 'Windows' and platform.python_implementation() == 'PyPy',
+        reason="Symlinks not supported on PyPy/Windows"
+    )
+    @pytest.mark.timeout(5)  # Set a 5-second timeout for this test
+    def test_folder_with_circular_symlink(self, tmp_path):
+
+        # Create a directory structure below with initial scannig point at tmp_path:
+        # tmp_path
+        # ├── dir
+        # │   └── file.txt
+        # └── symlink_dir -> dir
+
+        (tmp_path / "dir").mkdir()
+
+        (tmp_path / "dir" / "file1.txt").write_text("content1")
+
+        symlink_dir = tmp_path / "dir" / "symlink_dir"
+        symlink_dir.symlink_to(tmp_path / "dir")
+
+        folder = LocalFolder(str(tmp_path))
+
+        local_paths = folder.all_files(reporter=MagicMock())
+        absolute_paths = [path.absolute_path for path in list(local_paths)]
+
+        assert absolute_paths == [
+            fix_windows_path_limit(str(tmp_path / "dir" / "file1.txt")),
+            fix_windows_path_limit(str(tmp_path / "dir" / "symlink_dir" / "file1.txt")),
+        ]
+
+    @pytest.mark.skipif(
+        platform.system() == 'Windows' and platform.python_implementation() == 'PyPy',
+        reason="Symlinks not supported on PyPy/Windows"
+    )
+    @pytest.mark.timeout(5)  # Set a 5-second timeout for this test
+    def test_folder_with_symlink_to_parent(self, tmp_path):
+
```

Comment:
Some tests should be added:

1. test_root_short_loop where `tmp_path` is a symlink to itself
2. test_root_parent_loop where `tmp_path` is a symlink to its parent
3. test_symlink_that_points_deeper, `a, b/c -> b/d/e, f`
4. test_symlink_that_poitns_up, `a, b/c/d -> b/e, f`
5. test_elaborate_infinite_loop (symlink to a symlink to a symlink to a symlink which points to the first symlink), `a, b -> c, c -> d, d -> e, e -> b, f`
6. test_valid_symlink_pattern_where_the_link_goes_down_and_up `a, b -> c/d, d -> e, f`
7. test_valid_symlink_pattern_where_the_link_goes_up_and_down `a, b/c -> d, d -> e/f, g`
8. test_loop_that_goes_down_and_up, `a, b -> c/d, d -> e, e -> b, f`
9. test_loop_that_goes_up_and_down `a, b/c -> d, d -> e/f, f -> b/c, g`

I think your implementation is not vulnerable to any issues that this test might detect, but the tests are also here so that when someone tries to "optimize" your code in the future then any issues with the new implementation will be caught by the test harness.

Reply:
@ppolewicz what is the purpose of this test:

> test_root_parent_loop where tmp_path is a symlink to its parent

And what would you expect to see when running it?

Reply:
@ppolewicz The newest version contains all the tests above apart from the one I asked about earlier. It would be great if you could (at least) review the directory structures in the comments to make sure I understood your intentions correctly. In the mean time, I need to debug why the new tests fail on Windows 🤯🤯🤯

Reply:
it's a symlink like `/home/user/aaa` -> `/home/user` and then we try to sync `/home/user/aaa` somewhere